### ISSUE TITLE
Negative temperatures with radiation exchange model

### ISSUE THREAD

**Issue by PMehrfeld** [RWTH-EBC/AixLib#535]
## What is the problem?

- Solver calculates negative values when solving Stefan-Boltzmann equation in `AixLib.Utilities.HeatTransfer.HeatToStar`

## Why do we want to solve it?

- Although the solution is correct from a mathematical point of view, negative temperatures must be avoided.

## How do we want to solve it?

- ...

## Further description

I have a reproducible total model, in which the error occurs, when using `lsodar`. With `dassl` the simulation runs successfully.

@Mathadon and @rubenbaetens: I found your [old issue](https://github.com/open-ideas/IDEAS/issues/284), after I inspected the model `IDEAS.Buildings.Components.BaseClasses.RadiativeHeatTransfer.HeatRadiation`. By default `linearise = true` you use the linearized function instead of the biquadratic one. But have you ever encountered a problem like the screenshot shows when you used the biquadratic function? And if yes is there any solution for this problem?

Thanks a lot in advance for any help!

Screenshot:
![image](https://user-images.githubusercontent.com/11769433/36261083-856e2730-1263-11e8-80df-3d10fec48a05.png)

**Comment by Mathadon** [comment]
not sure what's going on here without more context :)

**Comment by PMehrfeld** [comment]
Of course, sorry, I'll explain it a bit more detailed.

When running an HVAC and building performance simulation, sometimes in the following equation of the [`HeatToStar` model](https://github.com/RWTH-EBC/AixLib/blob/master/AixLib/Utilities/HeatTransfer/HeatToStar.mo) the negativ solution for `Therm.T` is determined during the simulation (see screenshot above).

```Modelica
Therm.Q_flow = Modelica.Constants.sigma * eps * A * (Therm.T * Therm.T * Therm.T * Therm.T - Star.T * Star.T * Star.T * Star.T);
```

We use the `HeatToStar` model in order to determine the heat flows between wall surface temperatures and radation temperatures within a room. From my understanding this model should be similar to IDEAS' [`HeatRadiation`](https://github.com/open-ideas/IDEAS/blob/master/IDEAS/Buildings/Components/BaseClasses/RadiativeHeatTransfer/HeatRadiation.mo).

I was just wondering if you somehow encountered this problem as well with your simulations and the IDEAS library.

**Comment by Mathadon** [comment]
ok, now I get it :p 
As you said, the solution is probably mathematically correct, but it does not make sense from a physical point of view. A solution would be to use `max(Therm.T,0)` in your equation instead of `Therm.T`. This removes `-Therm.T` as a feasible variable. However, this introduces discontinuities in your equations.

With respect to how we deal with this: I suggest you read our publication (published today, lol) on how we implement radiative heat transfer :) 
http://www.tandfonline.com/doi/full/10.1080/19401493.2018.1428361

**Comment by PMehrfeld** [comment]
Thanks a lot @Mathadon. Actually I tried this before, since I also assumed that the negative solutions would be removed. But I got the same results. As you now ensured that with `max()` the negative solutions should really be omitted, I tried using `max(Therm.T,100)` and `max(Therm.T,100)`. Now the simulation works :)

I assume that maybe the solution 0 would lead to a division by 0 and, therefore, the solver "prefers" to break the max constraint...

Furthermore, I performed a quick check of simulation speed (using the `dassl` solver):

1. max(T,100): 100.00 %

2. T\*T\*T\*T: 98.05 %

3. T^4: 104.13 %


But for stability reasons I will of course change this in `development` branch as a bug fix to `max(T,100) * max(T,100) ...`.

**Comment by PMehrfeld** [comment]
And thanks a lot for referencing the brand new publication ;)
We will have a deeper look into it.

**Comment by rubenbaetens** [comment]
Fyi, remember that (A^4-B^4) can be reduced to (A-B)*(A+B)*(A^2+B^2), this is usually numerically more stable.

**Comment by PMehrfeld** [comment]
Thanks @rubenbaetens. This was also one of the solutions I tried prior opening this issue. However, this led to a different curve for the temperature, but also including parts with negative solutions.

Again thanks to both of you for the input.

**Comment by rubenbaetens** [comment]
I remember having a similar bug in a very early version of IDEAS. I think one of the solutions that worked was to put a state in Star.T ...

**Comment by PMehrfeld** [comment]
Closed by #536
